zome.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2013-2017, The MetaCurrency Project (Eric Harris-Braun, Arthur Brock, et. al.)
+// Use of this source code is governed by GPLv3 found in the LICENSE file
+//---------------------------------------------------------------------------------------
+
+// Data integrity engine for distributed applications -- a validating monotonic
+//
+// Zome struct encapsulates logically related code, from "chromosome"
+
+package holochain
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Zome struct {
+	Name        string
+	Description string
+	Code        string // file name of DNA code
+	CodeHash    Hash
+	Entries     map[string]EntryDef
+	NucleusType string
+	Functions   map[string]FunctionDef
+
+	// cache for code and nucleus
+	code string
+	nuc  *Nucleus
+}
+
+// EntryDef struct holds an entry definition
+type EntryDef struct {
+	Name       string
+	DataFormat string
+	Schema     string // file name of schema or language schema directive
+	SchemaHash Hash
+	Sharing    string
+	validator  SchemaValidator
+}
+
+// FunctionDef holds the name and calling type of an DNA exposed function
+type FunctionDef struct {
+	Name        string
+	CallingType string
+}
+
+// ZomePath returns the path to the zome dna data
+// @todo sanitize the name value
+func (h *Holochain) ZomePath(z *Zome) string {
+	return h.DNAPath() + "/" + z.Name
+}
+
+func (h *Holochain) PrepareZomes(zomes map[string]Zome) (err error) {
+	for _, z := range zomes {
+		zpath := h.ZomePath(&z)
+		if !fileExists(zpath + "/" + z.Code) {
+			fmt.Printf("%v", z)
+			err = errors.New("DNA specified code file missing: " + z.Code)
+			return
+		}
+		for name, e := range z.Entries {
+			sc := e.Schema
+			if sc != "" {
+				if !fileExists(zpath + "/" + sc) {
+					err = errors.New("DNA specified schema file missing: " + sc)
+					return
+				}
+				if strings.HasSuffix(sc, ".json") {
+					if err = e.BuildJSONSchemaValidator(zpath); err != nil {
+						return
+					}
+					z.Entries[name] = e
+				}
+			}
+		}
+	}
+
+	h.dht = NewDHT(h)
+
+	return
+}
+
+func (zome *Zome) GetNucleus(chain *Holochain) (err error) {
```

Comment:
This isn't quite complete. I want to store the nucleus on the zome, but I think I was coding a 2nd lookup map on the Holochain object. I want to avoid re-creating the zome on every call and then they will be ready to go for interzome calling as well.